btf/dedup.go

```diff
@@ -0,0 +1,457 @@
+	dedupMap := newDeduper()
+
+	uniqueRoots := make([]Type, 0)
+	seen := make(map[Type]struct{})
+	for _, root := range roots {
+		dedupedRoot := dedupMap.Deduplicate(root)
+		if _, ok := seen[dedupedRoot]; ok {
+			continue
+		}
+		seen[dedupedRoot] = struct{}{}
+		uniqueRoots = append(uniqueRoots, dedupedRoot)
+	}
+
+	return uniqueRoots, nil
+}
+
+type deduper struct {
+	visited    map[Type]struct{}
+	hashCache  map[hashCacheKey]uint64
+	known      map[Type]Type
+	candidates map[uint64][]Type
+	eqCache    map[typKey]bool
+}
+
+func newDeduper() *deduper {
+	return &deduper{
+		make(map[Type]struct{}),
+		make(map[hashCacheKey]uint64),
+		make(map[Type]Type),
+		make(map[uint64][]Type),
+		make(map[typKey]bool),
+	}
+}
+
+func (dm *deduper) Deduplicate(t Type) Type {
+	// If we have already attempted to deduplicate this exact type, return the result.
+	if deduped, ok := dm.known[t]; ok {
+		return deduped
+	}
+
+	// Visit the subtree, if a type has children, attempt to replace it with
+	// a deduplicated version of those children.
+	for t := range postorder(t, dm.visited) {
+		for c := range children(t) {
+			*c = dm.deduplicateSingle(*c)
+		}
+	}
+
+	// Finally, deduplicate the root type itself.
+	return dm.deduplicateSingle(t)
+}
+
+func (dm *deduper) deduplicateSingle(t Type) Type {
+	// If we have deduplicated this type before, return the result of that deduplication.
+	if deduped, ok := dm.known[t]; ok {
+		return deduped
+	}
+
+	// Compute the hash of this type. Types with the same hash are candidates for deduplication.
+	hash := typeHash(t, -1, dm.hashCache)
+
+	// A hash collision is possible, so we need to compare against all candidates with the same hash.
+	candidates := dm.candidates[hash]
+	for _, candidate := range candidates {
+		// Pre-size the visited slice, experimentation on VMLinux shows a capacity of 16 to give the best performance.
+		const visitedCapacity = 16
+		if typesEqual(candidate, t, make([]Type, 0, visitedCapacity), dm.eqCache) {
+			dm.known[t] = candidate
+			return candidate
+		}
+	}
+
+	dm.candidates[hash] = append(dm.candidates[hash], t)
+	return t
+}
+
+// maphash.Hash by default uses a random seed per Hash instance.
+// To ensure consistent hashing across multiple calls, we use a fixed seed for the lifetime of the program.
+var seed = maphash.MakeSeed()
+
+// The hash of a type is the same given the type and depth budget, and thus is the key for the cache.
+type hashCacheKey struct {
+	t           Type
+	depthBudget int
+}
+
+// typeHash computes a hash for `t`. The produced hash is the same for types which are similar.
+// The hash can collide such that two different types may produce the same hash, so equality must be checked separately.
+// It will recursively call itself to hash child types. The initial call should use a depthBudget of -1.
+func typeHash(t Type, depthBudget int, cache map[hashCacheKey]uint64) (sum uint64) {
+	if depthBudget == 0 {
+		return 0
```

Comment:
So this means that if a type is nested too deeply we use 0 as the hash? Seems unintuitive to give 0 that meaning, especially because it means that the 0 value then potentially has a large number of collisions in a pathological case.

Could this return an error instead?

If this needs to stay then the doc comment should mention this.